File: src/mcpbox/server/routes/servers.py
from datetime import datetime, timezone

from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse

from mcpbox.shared.config import Config
from mcpbox.shared.models import CreateServerRequest, UpdateServerRequest
from mcpbox.shared.s3 import (
    list_servers as s3_list_servers,
    get_server as s3_get_server,
    upsert_server,
    delete_server as s3_delete_server,
)

router = APIRouter()

_cfg = Config()
S3_BUCKET = _cfg.S3_BUCKET_NAME


@router.get("/{server_name}")
async def get_server(server_name: str) -> JSONResponse:
    """Get detailed information about a specific MCP server"""
    try:
        server = s3_get_server(S3_BUCKET, server_name)

        if not server:
            raise HTTPException(status_code=404, detail=f"Server '{server_name}' not found")

        return JSONResponse(content={"status": "success", "server": server})

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching server info: {str(e)}")


@router.get("")
async def list_servers() -> JSONResponse:
    """List all MCP servers from S3 bucket"""
    try:
        server_map = s3_list_servers(S3_BUCKET)

        if not server_map:
            return JSONResponse(content={"status": "success", "total": 0, "servers": []})

        server_list = []
        for server in server_map.values():
            server_info = {
                "name": server.get("name"),
                "version": server.get("version"),
                "description": server.get("description"),
                "author": server.get("author"),
                "lang": server.get("lang"),
                "license": server.get("license"),
                "entrypoint": server.get("entrypoint"),
                "repository": server.get("repository"),
            }

            if "tools" in server and server["tools"]:
                server_info["tools"] = server["tools"]

            if "pricing" in server and server["pricing"]:
                server_info["pricing"] = server["pricing"]
            else:
                server_info["pricing"] = {"currency": "", "amount": 0}

            if "security_report" in server and server["security_report"]:
                server_info["security_report"] = server["security_report"]

            server_list.append(server_info)

        return JSONResponse(
            content={"status": "success", "total": len(server_list), "servers": server_list}
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching servers: {str(e)}")


@router.post("")
async def create_server(server: CreateServerRequest) -> JSONResponse:
    """Create a new MCP server and add it to S3"""
    try:
        existing = s3_get_server(S3_BUCKET, server.name)
        if existing:
            raise HTTPException(status_code=400, detail=f"Server '{server.name}' already exists")

        new_server = {
            "name": server.name,
            "version": server.version,
            "description": server.description,
            "author": server.author,
            "lang": server.lang,
            "license": server.license,
            "entrypoint": server.entrypoint,
            "repository": {"type": server.repository.type, "url": server.repository.url},
            "meta": {
                "created_at": datetime.now(timezone.utc).isoformat(),
                "updated_at": datetime.now(timezone.utc).isoformat(),
            },
        }

        new_server["pricing"] = {
            "currency": server.pricing.currency,
            "amount": server.pricing.amount,
        }

        if server.tools:
            new_server["tools"] = server.tools

        upsert_server(S3_BUCKET, server.name, new_server)

        return JSONResponse(
            content={"status": "success", "message": "Server created", "server": new_server},
            status_code=201,
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error creating server: {str(e)}")


@router.put("/{server_name}")
async def update_server(server_name: str, updates: UpdateServerRequest) -> JSONResponse:
    """Update an existing MCP server with partial updates"""
    try:
        existing = s3_get_server(S3_BUCKET, server_name)
        if not existing:
            raise HTTPException(status_code=404, detail=f"Server '{server_name}' not found")

        updated_data = dict(existing)

        new_name = updates.name if updates.name else server_name
        if updates.name and updates.name != server_name:
            if s3_get_server(S3_BUCKET, updates.name):
                raise HTTPException(
                    status_code=400, detail=f"Server '{updates.name}' already exists"
                )
            updated_data["name"] = updates.name

        if updates.version is not None:
            updated_data["version"] = updates.version
        if updates.description is not None:
            updated_data["description"] = updates.description
        if updates.author is not None:
            updated_data["author"] = updates.author
        if updates.lang is not None:
            updated_data["lang"] = updates.lang
        if updates.license is not None:
            updated_data["license"] = updates.license
        if updates.entrypoint is not None:
            updated_data["entrypoint"] = updates.entrypoint
        if updates.repository is not None:
            updated_data["repository"] = {
                "type": updates.repository.type,
                "url": updates.repository.url,
            }
        if updates.pricing is not None:
            updated_data["pricing"] = {
                "currency": updates.pricing.currency,
                "amount": updates.pricing.amount,
            }
        if updates.tools is not None:
            updated_data["tools"] = updates.tools
        if updates.security_report is not None:
            updated_data["security_report"] = updates.security_report

        if "meta" in updated_data and "created_at" in updated_data["meta"]:
            del updated_data["meta"]["created_at"]

        if new_name != server_name:
            s3_delete_server(S3_BUCKET, server_name)

        upsert_server(S3_BUCKET, new_name, updated_data)

        return JSONResponse(
            content={
                "status": "success",
                "message": f"Server '{server_name}' updated successfully",
                "server": updated_data,
            }
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error updating server: {str(e)}")


@router.delete("/{server_name}")
async def delete_server(server_name: str) -> JSONResponse:
    """Delete an MCP server from the registry"""
    try:
        existing = s3_get_server(S3_BUCKET, server_name)
        if not existing:
            raise HTTPException(status_code=404, detail=f"Server '{server_name}' not found")

        success = s3_delete_server(S3_BUCKET, server_name)
        if not success:
            raise HTTPException(status_code=500, detail=f"Failed to delete server '{server_name}'")

        return JSONResponse(
            content={
                "status": "success",
                "message": f"Server '{server_name}' deleted successfully",
            }
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error deleting server: {str(e)}")

File: src/superbox/server/handlers/servers.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"superbox/server/models"

	"github.com/gin-gonic/gin"
)

func RegisterServers(api *gin.RouterGroup) {
	servers := api.Group("/servers")
	{
		servers.GET("", listServers)
		servers.GET("/:server_name", getServer)
		servers.POST("", createServer)
		servers.PUT("/:server_name", updateServer)
		servers.DELETE("/:server_name", deleteServer)
	}
}

func callPythonS3(function string, args map[string]interface{}) (map[string]interface{}, error) {
	scriptPath := filepath.Join("src", "superbox", "server", "helpers", "s3_helper.py")

	argsJSON, err := json.Marshal(map[string]interface{}{
		"function": function,
		"args":     args,
	})
	if err != nil {
		return nil, err
	}

	cmd := exec.Command("python", scriptPath, string(argsJSON))
	cmd.Env = os.Environ()
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("python s3 call failed: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(output, &result); err != nil {
		return nil, fmt.Errorf("failed to parse python output: %v", err)
	}

	if errMsg, ok := result["error"].(string); ok {
		return nil, fmt.Errorf("%s", errMsg)
	}

	return result, nil
}

func getServer(c *gin.Context) {
	serverName := c.Param("server_name")
	bucketName := os.Getenv("S3_BUCKET_NAME")

	result, err := callPythonS3("get_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": serverName,
	})
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"status": "error",
			"detail": "Server '" + serverName + "' not found",
		})
		return
	}

	server, ok := result["data"].(map[string]interface{})
	if !ok || server == nil {
		c.JSON(http.StatusNotFound, gin.H{
			"status": "error",
			"detail": "Server '" + serverName + "' not found",
		})
		return
	}

	c.JSON(http.StatusOK, models.ServerResponse{
		Status: "success",
		Server: server,
	})
}

func listServers(c *gin.Context) {
	bucketName := os.Getenv("S3_BUCKET_NAME")

	result, err := callPythonS3("list_servers", map[string]interface{}{
		"bucket_name": bucketName,
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": "error",
			"detail": "Error fetching servers: " + err.Error(),
		})
		return
	}

	serversMap, ok := result["data"].(map[string]interface{})
	if !ok {
		c.JSON(http.StatusOK, models.ServerResponse{
			Status:  "success",
			Total:   0,
			Servers: []interface{}{},
		})
		return
	}

	serverList := make([]interface{}, 0)
	for _, serverVal := range serversMap {
		server, ok := serverVal.(map[string]interface{})
		if !ok {
			continue
		}

		serverInfo := map[string]interface{}{
			"name":        server["name"],
			"version":     server["version"],
			"description": server["description"],
			"author":      server["author"],
			"lang":        server["lang"],
			"license":     server["license"],
			"entrypoint":  server["entrypoint"],
			"repository":  server["repository"],
		}

		if tools, ok := server["tools"].(map[string]interface{}); ok && tools != nil {
			serverInfo["tools"] = tools
		}

		if pricing, ok := server["pricing"].(map[string]interface{}); ok && pricing != nil {
			serverInfo["pricing"] = pricing
		} else {
			serverInfo["pricing"] = map[string]interface{}{
				"currency": "",
				"amount":   0,
			}
		}

		if securityReport, ok := server["security_report"].(map[string]interface{}); ok && securityReport != nil {
			serverInfo["security_report"] = securityReport
		}

		serverList = append(serverList, serverInfo)
	}

	c.JSON(http.StatusOK, models.ServerResponse{
		Status:  "success",
		Total:   len(serverList),
		Servers: serverList,
	})
}

func createServer(c *gin.Context) {
	var req models.CreateServerRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status": "error",
			"detail": "Invalid request: " + err.Error(),
		})
		return
	}

	bucketName := os.Getenv("S3_BUCKET_NAME")

	existing, err := callPythonS3("get_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": req.Name,
	})
	if err == nil && existing["data"] != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status": "error",
			"detail": "Server '" + req.Name + "' already exists",
		})
		return
	}

	newServer := map[string]interface{}{
		"name":        req.Name,
		"version":     req.Version,
		"description": req.Description,
		"author":      req.Author,
		"lang":        req.Lang,
		"license":     req.License,
		"entrypoint":  req.Entrypoint,
		"repository": map[string]interface{}{
			"type": req.Repository.Type,
			"url":  req.Repository.URL,
		},
		"pricing": map[string]interface{}{
			"currency": req.Pricing.Currency,
			"amount":   req.Pricing.Amount,
		},
		"meta": map[string]interface{}{
			"created_at": time.Now().UTC().Format(time.RFC3339),
			"updated_at": time.Now().UTC().Format(time.RFC3339),
		},
	}

	if req.Tools != nil {
		newServer["tools"] = *req.Tools
	}

	_, err = callPythonS3("upsert_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": req.Name,
		"server_data": newServer,
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": "error",
			"detail": "Error creating server: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusCreated, models.ServerResponse{
		Status:  "success",
		Message: "Server created",
		Server:  newServer,
	})
}

func updateServer(c *gin.Context) {
	serverName := c.Param("server_name")
	bucketName := os.Getenv("S3_BUCKET_NAME")

	var req models.UpdateServerRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status": "error",
			"detail": "Invalid request: " + err.Error(),
		})
		return
	}

	existingResult, err := callPythonS3("get_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": serverName,
	})
	if err != nil || existingResult["data"] == nil {
		c.JSON(http.StatusNotFound, gin.H{
			"status": "error",
			"detail": "Server '" + serverName + "' not found",
		})
		return
	}

	existing := existingResult["data"].(map[string]interface{})
	updatedData := make(map[string]interface{})
	for k, v := range existing {
		updatedData[k] = v
	}

	newName := serverName
	if req.Name != nil && *req.Name != serverName {
		checkResult, _ := callPythonS3("get_server", map[string]interface{}{
			"bucket_name": bucketName,
			"server_name": *req.Name,
		})
		if checkResult["data"] != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"status": "error",
				"detail": "Server '" + *req.Name + "' already exists",
			})
			return
		}
		newName = *req.Name
		updatedData["name"] = *req.Name
	}

	if req.Version != nil {
		updatedData["version"] = *req.Version
	}
	if req.Description != nil {
		updatedData["description"] = *req.Description
	}
	if req.Author != nil {
		updatedData["author"] = *req.Author
	}
	if req.Lang != nil {
		updatedData["lang"] = *req.Lang
	}
	if req.License != nil {
		updatedData["license"] = *req.License
	}
	if req.Entrypoint != nil {
		updatedData["entrypoint"] = *req.Entrypoint
	}
	if req.Repository != nil {
		updatedData["repository"] = map[string]interface{}{
			"type": req.Repository.Type,
			"url":  req.Repository.URL,
		}
	}
	if req.Pricing != nil {
		updatedData["pricing"] = map[string]interface{}{
			"currency": req.Pricing.Currency,
			"amount":   req.Pricing.Amount,
		}
	}
	if req.Tools != nil {
		updatedData["tools"] = *req.Tools
	}
	if req.SecurityReport != nil {
		updatedData["security_report"] = *req.SecurityReport
	}

	if meta, ok := updatedData["meta"].(map[string]interface{}); ok {
		if createdAt, exists := meta["created_at"]; exists {
			updatedData["meta"] = map[string]interface{}{
				"created_at": createdAt,
				"updated_at": time.Now().UTC().Format(time.RFC3339),
			}
		} else {
			updatedData["meta"] = map[string]interface{}{
				"updated_at": time.Now().UTC().Format(time.RFC3339),
			}
		}
	} else {
		updatedData["meta"] = map[string]interface{}{
			"updated_at": time.Now().UTC().Format(time.RFC3339),
		}
	}

	if newName != serverName {
		callPythonS3("delete_server", map[string]interface{}{
			"bucket_name": bucketName,
			"server_name": serverName,
		})
	}

	_, err = callPythonS3("upsert_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": newName,
		"server_data": updatedData,
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": "error",
			"detail": "Error updating server: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, models.ServerResponse{
		Status:  "success",
		Message: "Server '" + serverName + "' updated successfully",
		Server:  updatedData,
	})
}

func deleteServer(c *gin.Context) {
	serverName := c.Param("server_name")
	bucketName := os.Getenv("S3_BUCKET_NAME")

	existing, err := callPythonS3("get_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": serverName,
	})
	if err != nil || existing["data"] == nil {
		c.JSON(http.StatusNotFound, gin.H{
			"status": "error",
			"detail": "Server '" + serverName + "' not found",
		})
		return
	}

	_, err = callPythonS3("delete_server", map[string]interface{}{
		"bucket_name": bucketName,
		"server_name": serverName,
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status": "error",
			"detail": "Failed to delete server '" + serverName + "'",
		})
		return
	}

	c.JSON(http.StatusOK, models.ServerResponse{
		Status:  "success",
		Message: "Server '" + serverName + "' deleted successfully",
	})
}
